include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -138,18 +138,22 @@ internal_reinit(const Mapping<dim>                          &mapping,
       AssertDimension (dof_handler.size(), locally_owned_set.size());
 
       // set variables that are independent of FE
-      internal::assert_communicator_equality (dof_handler[0]->get_triangulation(),
-                                              additional_data.mpi_communicator);
-      size_info.communicator = additional_data.mpi_communicator;
       if (Utilities::MPI::job_supports_mpi() == true)
         {
+          const parallel::distributed::Triangulation<dim> *dist_tria =
+            dynamic_cast<const parallel::distributed::Triangulation<dim>*>
+            (&(dof_handler[0]->get_triangulation()));
+          size_info.communicator = dist_tria != 0 ?
+                                   dist_tria->get_communicator() :
+                                   MPI_COMM_SELF;
```

Comment:
`MatrixFree` should also work with `parallel::shared::Triangulation`, see `tests/matrix_free/matrix_vector_19`, so this cast should only check for `parallel::Triangulation`.